### DISCUSSION TITLE
How to build with github actions?

### DISCUSSION THREAD

**Discussion by kaeoticEve** [micropython/micropython#10672] [General]
due to some network problems, I get make errors every time I try to build locally, on WSL ubuntu 22.04 and on the esp-idf docker image with IDF v4.4. with both the latest branch and the 1.19.1 branch, even after I manually cloned all the submodules
as the main repository does utilize the GitHub actions, I've tried with the default YML config on my fork of the repo, it does finish building but the outcome is not getting outputted anywhere. are the actions only for validating?

Is there any way I can build a customized binary on the GitHub actions? 
the ideal configuration I'm attempting to create should contain ulab, microdot, and TensorFlow lite, but all these seem quite impossible after so many make errors.

**Comment by kaeoticEve** [comment]
ok, I figured it out eventually, the git submodules system is very different from python-pip or npm for js. 
the first step is the 'git clone recursive', and 'submodules init update', paste the submodule init update command a dozen times before you go to sleep, as this installation process only took 3 minutes on GitHub action took about 2 days to finish downloading on my network. if the command prints anything, the download isn't finished, you'll need to run it until it returns nothing.
many tutorials online consider this as the repositories are finished downloading, but sadly it is not the case, the git database might be ready but the actual code files are not. 
open the dir with vs code, and the git tab will tell you that the code base already has a bunch of modifications that await you to add and commit, they are all missing files git did not finish downloading, but the git thinks you have deleted them. so basically you'll need to restore those deleted files with the built-in discard changes button or use some git commands if you want to go the hard way.  after you clicked discard change on all repositories, you'll see another type of file that cannot be restored, they are the git diff files, they tell you that some submodules are not in the right branch/version/tag the root repository (in this case, esp-idf or micropython) requires. first, you need to manually switch branches for the root repo(esp-idf release-v4.4 and micropython master worked for me), the git diff files will tell you the required branch and the current branch of each miss-cloned repositories, and you'll have to git checkout all the repositories to the required branch, and git status on all repo to see if there are any unwanted changes. if there are, remove them, so your local repo can be identical to the online ones. luckily this step doesn't take long as the information git needed to switch branches is already downloaded, it's just tedious manual configurations that git should have automated.
after this you can follow the official installation guild to run the install.sh and the export.sh so the idf gets installed, and then you can run things as other tutorials to build the firmware afterward.

PS: to this date, the procedures above have produced a build environment that produces the bootloader, partition table, and application binary files, I'm going to test if they can be flashed to the board sometime later.

**Reply by mattytrentini** [reply]
You don't need to clone recursively, indeed that should be _avoided_ because many, _many_ submodules will be cloned - all of the code to build _all_ of the ports (and then some). MicroPython provides a `make submodules` command for each port which will clone just the appropriate repositories for the port you require. 

The steps to build with docker:

```bash
> git clone https://github.com/micropython/micropython.git
> cd micropython
> # Make your changes
> docker run -ti --rm -v $(pwd):/mp -w /mp espressif/idf:release-v4.4 bash -c "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=[YOURBOARD]"
```

It should take only a few minutes for the whole cycle. I just tested on a slowish laptop with a _wifi_ connection and it took about eight minutes from start to finish.

**Comment by kaeoticEve** [comment]
I've tried to flash the "firmware.bin" in the build folder to the esp32s3 via built-in USB (not UART) today, and it works ok, after some reconfigurations in the 'idf.py manuconfig', the firmware now recognizes all the 16m of flash and 8m of psram.
next step I'll try to mix other libs like lvgl, ulab, microdot, and TensorFlow lite in, I'll publish the build result on my fork if anything works
